src/miner/miner.rs

```diff
@@ -138,4 +141,56 @@ mod test {
 
     Ok(())
   }
+  #[test]
+  ///The worker should return a challenge with expected difficulty not realized difficulty.
+  fn worker_returns_challenge_with_expected_difficulty() -> Result<(), String> {
+    let miner = Miner::default();
+    let block_id: Vec<u8> = b"1111111111111111".iter().copied().collect();
+    let peer_id: Vec<u8> = b"1111111111111111".iter().copied().collect();
+
+    let timestamp: f64 = utc_seconds_f64();
+
+    let challenge: Challenge = Challenge {
+      difficulty: 2,
+      timestamp,
+      block_id: block_id.clone(),
+      peer_id: peer_id.clone(),
+    };
+
+    miner.worker.send(challenge.clone());
+    while None == miner.try_create_consensus() {}
+
+    let consensus: BlockConsensus;
+    //the second answer if any is guaranteed to be higher than the expected difficulty but the expected diff should stay the same
+    loop {
+      if let Some(new) = miner.try_create_consensus() {
+        consensus = BlockConsensus::deserialize(new.as_slice()).unwrap();
+        break;
+      };
+    }
+
+    assert_eq!(consensus.difficulty, challenge.difficulty);
+
+    let hash: H256 = mkhash(&mut get_hasher(), &block_id, &peer_id, consensus.nonce);
+
+    let (is_valid, realized_difficulty) = is_valid_proof_of_work(&hash, consensus.difficulty);
+    assert!(is_valid);
+    assert!(realized_difficulty > challenge.difficulty);
```

Comment:
I'm not sure what the above 3 lines are trying to accomplish. If the subject under test is `Worker::task()` which internally uses `is_valid_proof_of_work()` then inside the test we should not be calling directly `is_valid_proof_of_work()` and using its results unless I am missing something. 

But changing behavior in one function then calling another one to assert on its results doesn't look like the appropriate test. It looks like switching or mixing the subject under test.

Reply:
FTR: `is_valid_proof_of_work()` is a helper function which returns a tuple, bool and the leading number of zeroes in the solution (which we call the difficulty). 

Here it is used to calculate the difficulty of the returned solution (`hash` variable).